packages/model-fragments/lib/fragments/ext.js

```diff
@@ -152,83 +100,141 @@ Model.reopen({
+
+  Ember.keys(attrs).forEach(function(key) {
+    var attr = attrs[key];
+
+    // If the attribute has a `_createSnapshot` method, invoke it before the
+    // snapshot gets passed to the serializer
+    if (attr && typeof attr._createSnapshot === 'function') {
+      attrs[key] = attr._createSnapshot();
     }
-  },
+  });
 
-  /**
-    @method fragmentDidDirty
-    @private
-  */
-  fragmentDidDirty: function(key, fragment) {
-    if (!get(this, 'isDeleted')) {
-      // Add the fragment as a placeholder in the owner record's
-      // `_attributes` hash to indicate it is dirty
-      this._attributes[key] = fragment;
+  return snapshot;
+});
+
+/**
+  If the model `hasDirtyAttributes` this function will discard any unsaved
+  changes, recursively doing the same for all fragment properties.
+
+  Example
+
+  ```javascript
+  record.get('name'); // 'Untitled Document'
+  record.set('name', 'Doc 1');
+  record.get('name'); // 'Doc 1'
+  record.rollbackAttributes();
+  record.get('name'); // 'Untitled Document'
+  ```
 
-      this.send('becomeDirty');
+  @method rollbackAttributes
+*/
+decorateMethod(InternalModelPrototype, 'rollbackAttributes', function rollbackFragments() {
+  for (var key in this._fragments) {
+    if (this._fragments[key]) {
+      this._fragments[key].rollbackAttributes();
     }
-  },
+  }
+});
+
+/**
+  If the adapter did not return a hash in response to a commit,
+  merge the changed attributes and relationships into the existing
+  saved data and notify all fragments of the commit.
+
+  @method adapterDidCommit
+*/
+decorateMethod(InternalModelPrototype, 'adapterDidCommit', function adapterDidCommit(returnValue, args) {
+  var attributes = (args[0] && args[0].attributes) || {};
+  var fragment;
+
+  // Notify fragments that the record was committed
+  for (var key in this._fragments) {
+    if (fragment = this._fragments[key]) {
+      fragment._adapterDidCommit(attributes[key]);
+    }
+  }
+});
 
+/**
+  @class JSONSerializer
+  @namespace DS
+*/
+JSONSerializer.reopen({
   /**
-    @method fragmentDidReset
+    Enables fragment properties to have custom transforms based on the fragment
+    type, so that deserialization does not have to happen on the fly
+
+    @method transformFor
     @private
-    */
-  fragmentDidReset: function(key, fragment) {
-    // Make sure there's no entry in the owner record's
-    // `_attributes` hash to indicate the fragment is dirty
-    delete this._attributes[key];
-
-    // Don't reset if the record is new, otherwise it will enter the 'deleted' state
-    // NOTE: This case almost never happens with attributes because their initial value
-    // is always undefined, which is *usually* not what attributes get 'reset' to
-    if (!get(this, 'isNew')) {
-      this.send('propertyWasReset', key);
+  */
+  transformFor: function(attributeType) {
+    if (attributeType.indexOf('-mf-') === 0) {
+      return getFragmentTransform(this.container, this.store, attributeType);
```

Comment:
@igorT This is the other terrible thing I've done. This actually wasn't strictly necessary for compatibility with v1.13; it's just [something I've wanted to do for a while](https://github.com/lytics/ember-data.model-fragments/issues/12), and [keeps coming up](https://github.com/lytics/ember-data.model-fragments/issues/95). The gist is that when fragment attribute CPs access the data in `._data`, it should already be deserialized, which gives people an opportunity to override fragment transforms if they want/need.

Since the transform needs to know things about the specific fragment attribute (attribute type, fragment type, and polymorphic type key), and passing enough information to `Transform#deserialize` in [`applyTransforms`](https://github.com/emberjs/data/blob/v1.13.5/packages/ember-data/lib/serializers/json-serializer.js#L195) doesn't seem feasible, the only solution I could come up with is to create a new transform class. At first I balked at the thought, but then I came around, since there's very likely only a handful of transforms that are created in a given app.

The bad:

- This overrides a private method on `JSONSerializer`, which is never a good idea
- Relying on `JSONSerializer` breaks apps that have completely custom serializers (which I don't think is common, but still)
- Encoding type information as a string in the attribute type meta smells thoroughly. I'm not even sure if the `$` delimiter is intended to be a valid character in container keys